File: sdk/typescript/examples/twoWaySync.ts
/**
 * Two-way reconciliation end to end — via the `SyncEngine` facade.
 *
 * This example imports only the facade, policy types, and the adapter.
 * No `Orchestrator`, no `PolicyMap.withDefault(...)` ceremony — everything
 * internal stays internal.
 *
 * Run with:
 *   npx tsx examples/twoWaySync.ts
 */

import type { JsonValue } from "../src/domain/types.js";
import { idempotencyKey } from "../src/domain/idempotency.js";
import { TestMemoryAdapter } from "../src/adapters/testMemory.js";
import { Additive } from "../src/application/policy/additive.js";
import { OwnedBy } from "../src/application/policy/ownedBy.js";
import {
  StateMachine,
  StateTransition,
} from "../src/application/policy/stateMachine.js";
import { SyncEngine } from "../src/drivers/syncEngine.js";

const ENTITY = "purchase_order";
const PO_ID = "PO-42";

async function main(): Promise<void> {
  console.log("==== diff-fusion · two-way sync (facade) ====\n");

  // --------------------------------------------------------------------
  // Set up two systems. Users only touch the adapter type.
  // --------------------------------------------------------------------
  const erp = new TestMemoryAdapter("erp");
  const inv = new TestMemoryAdapter("inv");

  const starting: JsonValue = {
    price: 100,
    qty_recv: 5,
    status: "open",
  };
  erp.seed(ENTITY, PO_ID, starting);
  inv.seed(ENTITY, PO_ID, starting);

  // Simulate drift on both sides.
  await update(erp, { price: 120, qty_recv: 6, status: "closed" });
  await update(inv, { price: 999, qty_recv: 7, status: "closed" });

  console.log("BEFORE sync:");
  await printSide(erp, "  erp");
  await printSide(inv, "  inv");
  console.log();

  // --------------------------------------------------------------------
  // Build the engine. One chain of builder calls — no orchestrator
  // import, no internal types.
  // --------------------------------------------------------------------
  const engine = SyncEngine.builder(erp, inv)
    .policy("price", new OwnedBy("erp"))
    .policy("qty_recv", new Additive())
    .policy(
      "status",
      new StateMachine([
        new StateTransition("open", "closed"),
        new StateTransition("open", "cancelled"),
      ]),
    )
    .seedAncestor(ENTITY, PO_ID, starting)
    .build();

  // --------------------------------------------------------------------
  // Shadow run first (dry run) — the facade calls it `preview`.
  // --------------------------------------------------------------------
  const preview = await engine.preview(ENTITY, PO_ID);
  console.log("PREVIEW (no writes):");
  if (preview.wouldWrite !== undefined) {
    console.log(`  would write: ${JSON.stringify(preview.wouldWrite)}`);
  } else {
    console.log(
      `  would escalate — ${preview.conflicts.length} conflicts`,
    );
    for (const c of preview.conflicts) {
      console.log(`    · ${c.path} — ${c.reason}`);
    }
  }
  console.log();

  // --------------------------------------------------------------------
  // Real cycle.
  // --------------------------------------------------------------------
  const outcome = await engine.sync(ENTITY, PO_ID);
  switch (outcome.kind) {
    case "NoOp":
      console.log("Nothing to do.");
      break;
    case "Synced":
      console.log(`Synced. Pushed to: ${JSON.stringify(outcome.pushedTo)}`);
      break;
    case "Escalated":
      console.log(
        `Escalated — ${outcome.conflicts.length} conflict(s) queued:`,
      );
      for (const c of outcome.conflicts) {
        console.log(`  · ${c.path} — ${c.reason}`);
      }
      break;
    default: {
      const _exhaustive: never = outcome;
      throw new Error(`unreachable SyncOutcome: ${JSON.stringify(_exhaustive)}`);
    }
  }

  console.log("\nAFTER sync:");
  await printSide(erp, "  erp");
  await printSide(inv, "  inv");

  console.log(`\nEscalation queue depth: ${await engine.escalationDepth()}`);

  // Replay is a NoOp — ancestor advanced, no drift remains.
  const replay = await engine.sync(ENTITY, PO_ID);
  console.log(`Replay outcome: ${JSON.stringify(replay)}`);
}

/** Overwrite an adapter's current view (demo-only helper). */
async function update(
  port: TestMemoryAdapter,
  newValue: JsonValue,
): Promise<void> {
  const current = await port.findByCanonicalId(ENTITY, PO_ID);
  if (current === undefined) {
    throw new Error(`seeded entity ${ENTITY}/${PO_ID} not found`);
  }
  const ik = idempotencyKey(PO_ID, "upsert", newValue);
  await port.upsert(ENTITY, PO_ID, newValue, current.version, ik);
}

async function printSide(
  port: TestMemoryAdapter,
  label: string,
): Promise<void> {
  const ext = await port.findByCanonicalId(ENTITY, PO_ID);
  if (ext === undefined) throw new Error(`no entity on ${port.systemType()}`);
  const { canonical } = await port.fetch(ENTITY, ext);
  console.log(`${label} (${port.systemType()}): ${JSON.stringify(canonical)}`);
}

main().catch((e: unknown) => {
  console.error(e);
  process.exit(1);
});

File: core/examples/two_way_sync.rs
//! Two-way reconciliation end to end — via the `SyncEngine` facade.
//!
//! This example imports **only** the facade, the policy types, and the
//! adapter type. No `Orchestrator`, no `Arc<dyn …>`, no
//! `InMemoryAncestorStore`, no `PolicyMap::with_default(...)` ceremony.
//! Everything internal stays internal.
//!
//! Run with:
//!
//! ```bash
//! cargo run --example two_way_sync
//! ```

use diff_fusion::adapters::test_memory::TestMemoryAdapter;
use diff_fusion::domain::idempotency::idempotency_key;
use diff_fusion::application::policy::{Additive, OwnedBy, StateMachine, StateTransition};
use diff_fusion::ports::system::SystemPort;
use diff_fusion::CycleOutcome;
use diff_fusion::drivers::sync_engine::SyncEngine;
use serde_json::json;

const ENTITY: &str = "purchase_order";
const PO_ID: &str = "PO-42";

#[tokio::main(flavor = "current_thread")]
async fn main() {
    println!("==== diff-fusion · two-way sync (facade) ====\n");

    // ---------------------------------------------------------------
    // Set up two systems. Users only touch the adapter type.
    // ---------------------------------------------------------------
    let erp = TestMemoryAdapter::new("erp");
    let inv = TestMemoryAdapter::new("inv");

    let starting = json!({
        "price": 100,
        "qty_recv": 5,
        "status": "open",
    });
    erp.seed(ENTITY, PO_ID, starting.clone());
    inv.seed(ENTITY, PO_ID, starting.clone());

    // Simulate drift on both sides.
    update(&erp, json!({ "price": 120, "qty_recv": 6, "status": "closed" })).await;
    update(&inv, json!({ "price": 999, "qty_recv": 7, "status": "closed" })).await;

    println!("BEFORE sync:");
    print_side(&erp, "  erp").await;
    print_side(&inv, "  inv").await;
    println!();

    // ---------------------------------------------------------------
    // Build the engine. This is the user-facing API — one chain of
    // builder calls, no Arc wrapping, no orchestrator import.
    // ---------------------------------------------------------------
    let engine = SyncEngine::builder(erp.clone(), inv.clone())
        .policy("price", Box::new(OwnedBy::new("erp")))
        .policy("qty_recv", Box::new(Additive))
        .policy(
            "status",
            Box::new(StateMachine::new([
                StateTransition::new("open", "closed"),
                StateTransition::new("open", "cancelled"),
            ])),
        )
        .seed_ancestor(ENTITY, PO_ID, starting.clone())
        .build();

    // ---------------------------------------------------------------
    // Shadow run first (dry run) — the facade calls it `preview`.
    // ---------------------------------------------------------------
    let preview = engine.preview(ENTITY, PO_ID).await.unwrap();
    println!("PREVIEW (no writes):");
    if let Some(w) = &preview.would_write {
        println!("  would write: {w}");
    } else {
        let conflicts = &preview.resolution.conflicts;
        println!("  would escalate — {} conflicts", conflicts.len());
        for c in conflicts {
            println!("    · {} — {}", c.path, c.reason);
        }
    }
    println!();

    // ---------------------------------------------------------------
    // Real cycle.
    // ---------------------------------------------------------------
    let outcome = engine.sync(ENTITY, PO_ID).await.unwrap();
    match &outcome {
        CycleOutcome::NoOp => println!("Nothing to do."),
        CycleOutcome::Synced { pushed_to } => {
            println!("Synced. Pushed to: {pushed_to:?}");
        }
        CycleOutcome::Escalated { conflicts } => {
            println!("Escalated — {} conflict(s) queued:", conflicts.len());
            for c in conflicts {
                println!("  · {} — {}", c.path, c.reason);
            }
        }
    }

    println!("\nAFTER sync:");
    print_side(&erp, "  erp").await;
    print_side(&inv, "  inv").await;

    println!("\nEscalation queue depth: {}", engine.escalation_depth());

    // Replay is a NoOp — ancestor advanced, no drift remains.
    let replay = engine.sync(ENTITY, PO_ID).await.unwrap();
    println!("Replay outcome: {replay:?}");
}

/// Test-only helper: overwrite an adapter's current view.
async fn update(port: &TestMemoryAdapter, new_value: serde_json::Value) {
    let current = port
        .find_by_canonical_id(ENTITY, PO_ID)
        .await
        .unwrap()
        .expect("seeded entity");
    let ik = idempotency_key(PO_ID, "upsert", &new_value);
    port.upsert(ENTITY, PO_ID, &new_value, current.version.as_deref(), &ik)
        .await
        .unwrap();
}

async fn print_side(port: &TestMemoryAdapter, label: &str) {
    let ext = port.find_by_canonical_id(ENTITY, PO_ID).await.unwrap().unwrap();
    let (val, _) = port.fetch(ENTITY, &ext).await.unwrap();
    println!("{label} ({}): {val}", port.system_type());
}
